Tests/OptimizelyTests-MultiClients/NotificationCenterTests_MultiClients.swift

```diff
@@ -0,0 +1,250 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at   
+// 
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import XCTest
+
+class NotificationCenterTests_MultiClients: XCTestCase {
+    
+    let notificationCenter = DefaultNotificationCenter()
+    let lock = DispatchQueue(label: "notif-lock")
+    var counter = NotificationsCounter()
+
+    func testConcurrentSend() {
+        let numThreads = 10
+        let numEventsPerThread = 100
+        
+        counter = NotificationsCounter()
+        
+        _ = self.addTrackListener()
+        _ = self.addDecisionListener()
+        _ = self.addDatafileChangeListener()
+        _ = self.addLogEventListener()
+
+        let result = OTUtils.runConcurrent(count: numThreads) { thIdx in
+            (0..<numEventsPerThread).forEach { idx in
+                self.sendTrack()
+                self.sendDecision()
+                self.sendDatafileChange()
+                self.sendLogEvent()
+            }
+        }
+        
+        XCTAssertTrue(result, "Concurrent tasks timed out")
+        
+        self.lock.sync {}
+
+        XCTAssertEqual(counter.activate, 0)
+        XCTAssertEqual(counter.track, numThreads * numEventsPerThread)
+        XCTAssertEqual(counter.decision, numThreads * numEventsPerThread)
+        XCTAssertEqual(counter.datafileChange, numThreads * numEventsPerThread)
+        XCTAssertEqual(counter.logEvent, numThreads * numEventsPerThread)
+    }
+    
+    func testConcurrentAdd() {
+        let numThreads = 10
+        let numEventsPerThread = 100
+        
+        counter = NotificationsCounter()
+        
+        let result = OTUtils.runConcurrent(count: numThreads) { thIdx in
+            (0..<numEventsPerThread).forEach { idx in
+                _ = self.addTrackListener()
+                _ = self.addDecisionListener()
+                _ = self.addDatafileChangeListener()
+                _ = self.addLogEventListener()
+            }
+        }
+        
+        XCTAssertTrue(result, "Concurrent tasks timed out")
+        
+        self.sendTrack()
+        self.sendDecision()
+        self.sendDatafileChange()
+        self.sendLogEvent()
+
+        self.lock.sync {}
+
+        XCTAssertEqual(counter.activate, 0)
+        XCTAssertEqual(counter.track, numThreads * numEventsPerThread)
+        XCTAssertEqual(counter.decision, numThreads * numEventsPerThread)
+        XCTAssertEqual(counter.datafileChange, numThreads * numEventsPerThread)
+        XCTAssertEqual(counter.logEvent, numThreads * numEventsPerThread)
+    }
+    
+    func testConcurrentAddRemove() {
```

Comment:
can we update this test case to remove only a part of the notification listeners. For example if we have 10 log event listeners, just remove 1 or 2 of them? because the current implementation seems very similar to `testConcurrentAddClear`. This is help us assert the remaining unremoved listeners of the similar kind.